torch/csrc/jit/passes/symbolic_shape_analysis.cpp

```diff
@@ -86,54 +103,105 @@ struct SymbolicShapeAnalyzer {
+      // XXX: we cannot substitute symbolic dims before passes like constant
+      // propagation, or we might inadvertently use them in arithmetic or
+      // other operators
+      substituteInputTensorProperties(/*substitute_symbolic_dims*/ false);
       LowerSimpleTuples(graph_);
       RemoveListMutation(graph_);
       UnrollConstantLoops(graph_);
       ConstantPropagation(graph_);
       PeepholeOptimize(graph_);
       ConstantPropagation(graph_);
+      EliminateCommonSubexpression(graph_);
     }
-    ConstantPooling(graph_);
-    EliminateDeadCode(graph_);
+    substituteInputTensorProperties(/*substitute_symbolic_dims*/ true);
+    // XXX: do not run any passes aftr we have substituted in symbolic dimension
+    // value, we do it so they can be easily extracted into the output shape
     return extractOutputShape();
   }
 
  private:
-  void substituteInputTensorProperties() {
-    for (auto index : node_input_tensor_indices) {
-      substituteTensorProperties(index);
+  void substituteInputTensorProperties(bool substitute_symbolic_dims) {
+    // here we iteratively substitute properties of the node's input tensors
+    // into the shape compute graph. in addition to direct constants we can
+    // substitute, like len(inp) or inp[0] if the tensor has fixed length
+    // or first dimension, we also try to resolve symbolic shapes of the same
+    // symbolic value to the same Value * in the shape compute graph.
+    // for the shape logic:
+    // dim1 = inp1[0], dim2 = inp2[0]; return dim1 if dim2 == 1 else dim2
+    // if we see that inp1[0] and inp2[0] both have the same symbolic shape
+    // value, then it is a valid transformation to replace dim2 with dim1 or
+    // vice versa. to do this we collect  all Value * for a particular symbolic
+    // dimension value and then Value * with their dominator of the same
+    // symbolic dimension value in the example above, this allows us to infer
+    // that the output will be the symbolic dimension value of dim1
+
+    std::unordered_map<int64_t, std::vector<Value*>> symbolic_shape_map;
+
+    for (auto tensor_index : node_input_tensor_indices) {
+      auto tensor_value = node_->inputs().at(tensor_index);
+      auto tensor_shape =
+          tensor_value->type()->expect<TensorType>()->symbolic_sizes();
+      TORCH_INTERNAL_ASSERT(tensor_shape.rank().has_value());
+
+      for (const auto& use : graph_->inputs().at(tensor_index)->uses()) {
+        // TODO: either decompose composite ops like slice or add handling here
+        switch (use.user->kind()) {
+          case aten::len: {
+            size_t len = tensor_shape.rank().value();
+            replaceWithIValue(use.user->output(), static_cast<int64_t>(len));
+          } break;
+          case aten::__getitem__: {
+            auto index = constant_as<int64_t>(use.user->inputs().at(1));
+            if (index) {
+              auto norm_index = normIndex(*index, *tensor_shape.rank());
+              if (norm_index &&
```

Comment:
Since both conditions have `norm_index`, it may be more readable to hoist `if (norm_index)` out as a top-level branch